### PR TITLE
[1/5] cluster: implement proper builder pattern & minor refactor

### PR DESCRIPTION
**Pull request by utkarshgupta137**
Most of the crates in the Rust ecosystem use `builder()` & `build()` for client & builder as well as `new()` for both of them. Currently, the redis crate uses the unintuitive `open()` methods for the same.
With this PR, I've changed the cluster client to follow the proper builder pattern and redirect the old methods to the new ones.

Also refactored the cluster client in 2 main ways:
- Moved the `build()` method from the client to the builder & made it public
- Create a crate-level struct `ClusterParams` to hold all the params inside a single struct instead of multiple fields in the main struct
  - This struct will be used in the later PRs to pass cluster params to & inside the ClusterConnection impl

In addition, improved the documentation as I saw fit.

### CODE REVIEW

File: redis/src/cluster_client.rs [hunk 1]
```diff
@@ -1,164 +1,186 @@
...
-    pub fn open(self) -> RedisResult<ClusterClient> {
-        ClusterClient::build(self)
+    /// Upon failure to parse initial nodes or if the initial nodes have different passwords or
+    /// usernames, an error is returned.
+    pub fn build(self) -> RedisResult<ClusterClient> {
+        let mut initial_nodes = self.initial_nodes?;
+
+        let mut cluster_params = self.cluster_params;
+        let password = if cluster_params.password.is_none() {
+            cluster_params.password = initial_nodes[0].redis.password.clone();
```
Comment (jaymell) [inline redis/src/cluster_client.rs:50]: This will cause a panic if `initial_nodes` is empty for some reason.

File: redis/src/cluster.rs [hunk 2]
```diff
@@ -95,25 +96,23 @@ impl TlsMode {
 
 impl ClusterConnection {
     pub(crate) fn new(
-        initial_nodes: Vec<ConnectionInfo>,
-        read_from_replicas: bool,
-        username: Option<String>,
-        password: Option<String>,
+        cluster_params: &ClusterParams,
```
Comment (jaymell) [inline redis/src/cluster.rs:99]: I'm not sure there's a reason to use references for the params here, given that you're just cloning anyway. I would let the caller do the cloning if needed.

Comment (utkarshgupta137) [reply redis/src/cluster.rs:99]: I prefer to pass references as much as possible so that if I don't need a clone later on, I don't have to modify the whole call chain. Also if I only need a part of the struct, I don't have to clone the whole struct.
This is a matter of personal opinion. If you prefer, I can modify the code to pass cloned struct.

**Review by jaymell — APPROVED** [review]

File: redis/src/cluster_client.rs [hunk 3]
```diff
@@ -44,6 +44,12 @@ impl ClusterClientBuilder {
     /// usernames, an error is returned.
     pub fn build(self) -> RedisResult<ClusterClient> {
         let mut initial_nodes = self.initial_nodes?;
+        if initial_nodes.is_empty() {
```
Comment (djc) [inline redis/src/cluster_client.rs:47]: Instead of doing it this way, I'd prefer to anchor this in the type system, something like this:

```rust
let first_node = match initial_nodes.first() {
    Some(node) => node,
    None => return Err(..),
};
```

then use `first_node` instead of `initial_nodes[0]` below.

File: redis/src/cluster_client.rs [hunk 4]
```diff
@@ -41,69 +43,72 @@ impl ClusterClientBuilder {
...
+        } else {
+            &None
+        };
 
-        for (index, info) in initial_nodes.into_iter().enumerate() {
-            if let ConnectionAddr::Unix(_) = info.addr {
-                return Err(RedisError::from((ErrorKind::InvalidClientConfig,
-                                         "This library cannot use unix socket because Redis's cluster command returns only cluster's IP and port.")));
-            }
+        initial_nodes = initial_nodes
```
Comment (djc) [inline redis/src/cluster_client.rs:68]: What's the impetus here for moving from a `for`-loop to iterator combinators? IMO in this case it doesn't much improve the code (and this is another change that is unrelated to the stated goal of the commit) -- in fact, it looks like it adds another level of indentation which makes the code harder to read.

Comment (utkarshgupta137) [reply redis/src/cluster_client.rs:69]: I initially thought that this function is called each time we need a new connection, so I changed it to an iterator pattern which would be (slightly) faster.

**Comment by utkarshgupta137** [comment]
@jaymell @djc It's failing on older versions due to an unrelated crate error:
```
error: package `async-global-executor v2.3.0` cannot be built because it requires rustc 1.59 or newer, while the currently active rustc version is 1.57.0
```

File: redis/src/cluster_client.rs [hunk 5]
```diff
@@ -104,6 +104,12 @@ impl ClusterClientBuilder {
         self
     }
 
+    /// Use `build()`.
```
Comment (djc) [inline redis/src/cluster_client.rs:122]: This should be in the same place as the previous `open()` method so the diff is more clear.

File: redis/src/cluster_client.rs [hunk 6]
```diff
@@ -147,11 +159,17 @@ impl ClusterClient {
             self.password.clone(),
         )
     }
+
+    /// Use `new()`.
+    #[deprecated(since = "0.22.0", note = "Use new()")]
+    pub fn open<T: IntoConnectionInfo>(initial_nodes: Vec<T>) -> RedisResult<ClusterClient> {
```
Comment (djc) [inline redis/src/cluster_client.rs:174]: Please keep this in the same place as the previous `open()` method at least for this commit, so it's obvious what changed.

File: redis/src/cluster.rs [hunk 7]
```diff
@@ -96,11 +96,11 @@ impl TlsMode {
 
 impl ClusterConnection {
     pub(crate) fn new(
-        cluster_params: &ClusterParams,
-        initial_nodes: &[ConnectionInfo],
+        cluster_params: ClusterParams,
```
Comment (djc) [inline redis/src/cluster.rs:99]: These changes should be part of the "use ClusterParams struct to pass params" commit.

File: redis/src/cluster_client.rs (hunk 3 above)
Comment (djc) [reply redis/src/cluster_client.rs]: You apparently applied these changes in some other place in an earlier commit, but not here.

**Comment by djc** [comment]
This is much better! A few more minor changes. Please add a separate commit that bumps the MSRV to 1.59.

File: redis/src/cluster_client.rs (hunk 4 above)
Comment (djc) [reply redis/src/cluster_client.rs:69]: How/why do you think it would be faster? I don't think that's true.

File: redis/src/cluster_client.rs (hunk 5 above)
Comment (utkarshgupta137) [reply redis/src/cluster_client.rs:122]: I've deliberately moved deprecated methods to the bottom of the struct to keep them out of the way.

File: redis/src/cluster_client.rs (hunk 6 above)
Comment (utkarshgupta137) [reply redis/src/cluster_client.rs:174]: I've deliberately moved deprecated methods to the bottom of the struct to keep them out of the way.

File: redis/src/cluster_client.rs (hunk 4 above)
Comment (utkarshgupta137) [reply redis/src/cluster_client.rs:69]: I'm relatively new to rust, but from what I've read, iterators can be parallelised better than for-loops & also we don't have to check & increase the capacity of the vector that we collect into. It is a very slight improvement.

Ref: https://users.rust-lang.org/t/we-all-know-iter-is-faster-than-loop-but-why/51486

**Comment by utkarshgupta137** [comment]
> This is much better! A few more minor changes. Please add a separate commit that bumps the MSRV to 1.59.

MSRV change is entirely unrelated to this PR. I've created a separate PR: https://github.com/redis-rs/redis-rs/pull/673

File: redis/src/cluster_client.rs (hunk 5 above)
Comment (djc) [reply redis/src/cluster_client.rs:122]: I'd prefer that that is done in a separate commit to faciliate review.

File: redis/src/cluster_client.rs (hunk 6 above)
Comment (djc) [reply redis/src/cluster_client.rs:174]: I'd prefer that that is done in a separate commit to faciliate review.

**Comment by utkarshgupta137** [comment]
@djc I've added the bump MSRV commit to this PR for testing purposes, but please merge #673 first.

File: redis/src/cluster_client.rs (hunk 4 above)
Comment (djc) [reply redis/src/cluster_client.rs:69]: Okay, please don't resolve my review comments unless you've done as I suggested.

There will be no meaningful difference in performance here, especially if you allocate the new `Vec` by using `with_capacity()` up front. The `for`-loop is more readable in this case.

File: redis/src/cluster_client.rs [hunk 8]
```diff
@@ -39,68 +41,69 @@ impl ClusterClientBuilder {
     /// usernames, an error is returned.
     pub fn build(self) -> RedisResult<ClusterClient> {
         let initial_nodes = self.initial_nodes?;
-
         let mut nodes = Vec::with_capacity(initial_nodes.len());
```
Comment (djc) [inline redis/src/cluster_client.rs:44]: Let's move this to directly before the loop.

File: redis/src/cluster_client.rs (hunk 1 above)
Comment (djc) [reply redis/src/cluster_client.rs]: So this invariant is not checked here. Is it checked somewhere else? If yes, then we should add a comment pointing that out. If not, should check it here.

Comment (djc) [reply redis/src/cluster_client.rs]: Ok.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Okay, so please rebase and self-review part 2 to have small commits and align with the other feedback I've given so far, this has taken a lot of time.
